plugin/documents.py

```diff
@@ -606,46 +614,70 @@ def render_highlights_on_main_thread() -> None:
 
     # --- textDocument/complete ----------------------------------------------------------------------------------------
 
-    def _on_query_completions_async(self, resolve: ResolveCompletionsFn, location: int) -> None:
-        session = self.session('completionProvider', location)
-        if not session:
-            resolve([], 0)
+    def _on_query_completions_async(self, resolve_completion_list: ResolveCompletionsFn, location: int) -> None:
+        sessions = self.sessions('completionProvider')
+        if not sessions:
+            resolve_completion_list([], 0)
             return
         self.purge_changes_async()
-        can_resolve_completion_items = bool(session.get_capability('completionProvider.resolveProvider'))
-        config_name = session.config.name
-        session.send_request_async(
-            Request.complete(text_document_position_params(self.view, location), self.view),
-            lambda res: self._on_complete_result(res, resolve, can_resolve_completion_items, config_name),
-            lambda res: self._on_complete_error(res, resolve))
-
-    def _on_complete_result(self, response: Optional[Union[dict, List]], resolve: ResolveCompletionsFn,
-                            can_resolve_completion_items: bool, session_name: str) -> None:
-        response_items = []  # type: List[Dict]
-        flags = 0
+
+        completion_promises = []  # type: List[Promise[CompletionResponseWithSession]]
+        for session in sessions:
+
+            def completion_request() -> Promise:
+                return Promise(lambda resolve: session.send_request_async(
+                    Request.complete(text_document_position_params(self.view, location), self.view),
+                    lambda res: self._on_complete_result(res, resolve, session),
+                    lambda res: self._on_complete_error(res, resolve, session)))
```

Comment:
Passing the session is a bit unsafe since it could in theory die during request.
You could resolve `can_resolve_completion_items` right away here.
But then you would also need to pass the session name as it's needed in `format_completion` later.

I'm fine with being overridden here as it's probably a fairly low chance of this being a problem and we are probably doing that in other places already.

Reply:
I will try to pass only the session_name, and later retrieve the session by the session_name.
I would want to pass as little parameters as possible :)

Reply:
64e3e29532f5bf8c3bab28457de52fd4ca023273